packages/juggler/src/model.ts

```diff
@@ -0,0 +1,140 @@
+// Copyright IBM Corp. 2017. All Rights Reserved.
+// Node module: juggler
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Options, AnyType} from './common';
+import {Type} from './types';
+
+/**
+ * This module defines the key classes representing building blocks for Domain
+ * Driven Design.
+ * See https://en.wikipedia.org/wiki/Domain-driven_design#Building_blocks
+ */
+
+/**
+ * Property definition for a model
+ */
+export class ModelProperty {
+  name: string;
+  type: string | Function | Object | Type<AnyType>; // For example, 'string', String, or {}
+  json?: PropertyForm;
+  store?: PropertyForm;
+  [attribute: string]: AnyType; // Other attributes
+}
+
+/**
+ * See https://github.com/strongloop/loopback-datasource-juggler/issues/432
+ */
+export interface PropertyForm {
+  in?: boolean; // Can the property be used for input
+  out?: boolean; // Can the property be used for output
+  name?: string; // Custom name for this form
+}
+
+/**
+ * Definition for a model
+ */
+export class ModelDefinition {
+  name: string;
+  properties: Map<string, ModelProperty>;
+  // indexes: Map<string, AnyType>;
+  [attribute: string]: AnyType; // Other attributes
+
+  idProperties(): ModelProperty[] {
+    return [];
+  }
+}
+
+/**
+ * Base class for models
+ */
+export abstract class Model {
+  static modelName: string;
+  static definition: ModelDefinition;
+
+  /**
+   * Serialize into a plain JSON object
+   */
+  toJSON(): Object {
+    return {};
+  }
+
+  /**
+   * Convert to a plain object as DTO
+   */
+  toObject(options?: Options): Object {
+    return {};
+  }
+
+  [prop: string]: AnyType;
+}
+
+export interface Persistable {
+  // isNew: boolean;
+}
+
+/**
+ * Base class for value objects - An object that contains attributes but has no
+ * conceptual identity. They should be treated as immutable.
+ */
+export abstract class ValueObject extends Model implements Persistable {
+}
+
+/**
+ * Base class for entities which have unique ids
+ */
+export abstract class Entity extends Model implements Persistable {
+  /**
+   * Get the identity value. If the identity is a composite key, returns
+   * an object.
```

Comment:
I think this won't work well with non-primitive ID values like MongoDB's `ObjectID`. How can the caller of `getId()` distinguish between the following two cases?
  - `getId()` returned an object with all composite-key values
  - `getId()` returned an ObjectID instance as the single ID used by the model

From my experience so far, composite keys are pretty broken in the current juggler. I am proposing to remove support of composite keys in this initial version of loopback-next, and then later research how to implement composite keys properly. I personally prefer to have a smaller feature set where each feature works really well, rather than a large set of features that may or may not work.


cc @ritch @superkhau